### PR TITLE
Update serializer decoder to support access to underlying buffer data address and size

### PR DESCRIPTION
**Pull request by aggarw13**
*Issue #, if available:*

*Description of changes:*
Add APIs to **serializer** abstraction for the following features:
 * Obtain the starting address of `IotSerializerDecoderObject_t` and `IotSerializerIterator_t` objects of data in encoded data buffer.
 * Obtain the length of the underlying encoded data represented by `IotSerializerDecoderObject_t` and `IotSerializerIterator_t` objects

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: libraries/standard/serializer/include/iot_serializer.h [hunk 1]
```diff
@@ -275,12 +289,14 @@ typedef struct IotSerializerEncoderObject
 typedef struct IotSerializerDecoderObject
 {
     IotSerializerDataType_t type;
+    const uint8_t * bufferPtr;
```
Comment (qiutongs) [inline libraries/standard/serializer/include/iot_serializer.h]: With our naming convention, it should be pBuffer

File: libraries/standard/serializer/include/iot_serializer.h [hunk 2]
```diff
@@ -275,12 +289,14 @@ typedef struct IotSerializerEncoderObject
...
+    size_t dataLengthInBuffer;
     union
     {
-        /* useful when this is a container */
+        /* Useful if the type is a container. */
         void * pHandle;
         /* if the type is a container, the scalarValue is unuseful */
-        IotSerializerScalarValue_t value;
+        IotSerializerScalarValue_t scalarValue;
```
Comment (qiutongs) [inline libraries/standard/serializer/include/iot_serializer.h]: This may break existing code. Be careful.

Comment (aggarw13) [reply libraries/standard/serializer/include/iot_serializer.h]: Good catch, I will revert that name change

File: libraries/standard/serializer/src/cbor/iot_serializer_tinycbor_decoder.c [hunk 3]
```diff
@@ -78,6 +103,22 @@ typedef struct _cborValueWrapper
 
 /*-----------------------------------------------------------*/
 
+static size_t calculateSizeOfCborObject( CborValue * pValue )
+{
+    IotSerializer_Assert( pValue != NULL );
+    CborValue nextValue;
+
+    /* Copy the contents of the current CBOR object so that we can advance to the next object. */
+    memcpy( &nextValue, pValue, sizeof( CborValue ) );
```
Comment (qiutongs) [inline libraries/standard/serializer/src/cbor/iot_serializer_tinycbor_decoder.c]: It is same as below, right?
```
nextValue = *pValue;
```

File: libraries/standard/serializer/src/cbor/iot_serializer_tinycbor_decoder.c [hunk 4]
```diff
@@ -78,6 +103,22 @@ typedef struct _cborValueWrapper
...
+
+    IotSerializer_Assert( cbor_value_advance( &nextValue ) == IOT_SERIALIZER_SUCCESS );
```
Comment (qiutongs) [inline libraries/standard/serializer/src/cbor/iot_serializer_tinycbor_decoder.c]: Better to make it separate line for easy debugging.

Comment (aggarw13) [reply libraries/standard/serializer/src/cbor/iot_serializer_tinycbor_decoder.c]: Good point. Will do so.

File: libraries/standard/serializer/src/cbor/iot_serializer_tinycbor_decoder.c (hunk 3 above)
Comment (aggarw13) [reply libraries/standard/serializer/src/cbor/iot_serializer_tinycbor_decoder.c]: I'm not sure whether C will allow **assignment-copy** of `CborValue` type in a compiler-agnostic way...

File: libraries/standard/serializer/src/cbor/iot_serializer_tinycbor_decoder.c [hunk 5]
```diff
@@ -467,6 +525,15 @@ static IotSerializerError_t _next( IotSerializerDecoderIterator_t iterator )
 
     cborError = cbor_value_advance( &pCborValueWrapper->cborValue );
 
+    if( cbor_value_at_end( &pCborValueWrapper->cborValue ) == false )
+    {
+        /* Update the iterator to store the encoded data pointer and length for the new element being pointed to. */
+        IotSerializerDecoderObject_t * pDecoderObject = ( IotSerializerDecoderObject_t * ) iterator;
+        pDecoderObject->bufferPtr = pCborValueWrapper->cborValue.ptr;
+        pDecoderObject->dataLengthInBuffer = calculateSizeOfCborObject(
```
Comment (ravibhagavandas) [inline libraries/standard/serializer/src/cbor/iot_serializer_tinycbor_decoder.c]: cbor_value_advance() is a costly operation for nested structures. Can we do it on demand from the API.

File: libraries/standard/serializer/src/cbor/iot_serializer_tinycbor_decoder.c (hunk 3 above)
Comment (qiutongs) [reply libraries/standard/serializer/src/cbor/iot_serializer_tinycbor_decoder.c]: Not have a reference to C standard. But I am pretty sure that is the case. You are safe to use it. (This is not  a PR-blocking change.)

File: libraries/standard/serializer/test/unit/iot_tests_serializer_cbor.c [hunk 6]
```diff
@@ -422,3 +430,223 @@ TEST( Serializer_Unit_CBOR, Encoder_map_nest_array )
...
+
+TEST_TEAR_DOWN( Serializer_Decoder_Unit_CBOR )
+{
+    IotSdk_Cleanup();
+}
+
+TEST_GROUP_RUNNER( Serializer_Decoder_Unit_CBOR )
+{
+    RUN_TEST_CASE( Serializer_Decoder_Unit_CBOR, TestDecoderObjectWithNestedMap );
+    /* RUN_TEST_CASE( Serializer_Decoder_Unit_CBOR, TestDecoderIteratorWithNestedMap ); */
```
Comment (qiutongs) [inline libraries/standard/serializer/test/unit/iot_tests_serializer_cbor.c]: Want to comment back?

File: libraries/standard/serializer/test/unit/iot_tests_serializer_cbor.c [hunk 7]
```diff
@@ -422,3 +430,223 @@ TEST( Serializer_Unit_CBOR, Encoder_map_nest_array )
...
+}
+
+TEST( Serializer_Decoder_Unit_CBOR, TestDecoderObjectWithNestedMap )
```
Comment (qiutongs) [inline libraries/standard/serializer/test/unit/iot_tests_serializer_cbor.c]: This test looks good to me.

File: libraries/standard/serializer/test/unit/iot_tests_serializer_cbor.c [hunk 8]
```diff
@@ -480,166 +481,172 @@ TEST( Serializer_Decoder_Unit_CBOR, TestDecoderObjectWithNestedMap )
...
     TEST_ASSERT_EQUAL( IOT_SERIALIZER_SUCCESS, _pCborDecoder->find( &outerMapDecoder1, "A",
                                                                     &innerMapDecoder ) );
     TEST_ASSERT_EQUAL( IOT_SERIALIZER_SCALAR_SIGNED_INT, innerMapDecoder.type );
 
-    /* Make sure that the getBufferLocationOfDecoderObject() API returns the first location in the buffer to the value
-     * for the nested entry keyed by
-     * "A".*/
-    TEST_ASSERT_EQUAL_PTR( &_testEncodedNestedMap[ 6 ],
-                           _pCborDecoder->getBufferLocationOfDecoderObject( &innerMapDecoder ) );
+    /* Make sure that the getBufferAddress() API does not support getting buffeer address of value in the
```
Comment (ravibhagavandas) [inline libraries/standard/serializer/test/unit/iot_tests_serializer_cbor.c]: typo 'buffer'

File: libraries/standard/serializer/test/unit/iot_tests_serializer_cbor.c [hunk 9]
```diff
@@ -480,166 +481,172 @@ TEST( Serializer_Decoder_Unit_CBOR, TestDecoderObjectWithNestedMap )
...
-
-    /* Iterate to the end of the nested map container. */
-    TEST_ASSERT_EQUAL( IOT_SERIALIZER_SUCCESS, _pCborDecoder->next( nestedMapIter ) );
-
-    TEST_ASSERT_EQUAL( IOT_SERIALIZER_SUCCESS, _pCborDecoder->stepOut( nestedMapIter,
-                                                                       &nestedMapDecoder ) );
-    _pCborDecoder->destroy( &nestedMapDecoder );
-    _pCborDecoder->destroy( &outerDecoder2 );
-}
+/* TEST( Serializer_Decoder_Unit_CBOR, TestDecoderIteratorWithNestedMap ) */
```
Comment (ravibhagavandas) [inline libraries/standard/serializer/test/unit/iot_tests_serializer_cbor.c]: Un-comment test case ?

**Comment by abhidixi11** [comment]
Please make sure to run tests and demo for defender since defender library also uses serializer.

**Comment by qiutongs** [comment]
> Please make sure to run tests and demo for defender since defender library also uses serializer.

Just to add, defender library itself only uses encoder of serializer but defender tests uses decoder of serialzier.

File: libraries/standard/serializer/test/unit/iot_tests_serializer_cbor.c (hunk 9 above)
Comment (aggarw13) [reply libraries/standard/serializer/test/unit/iot_tests_serializer_cbor.c]: Done

File: libraries/standard/serializer/test/unit/iot_tests_serializer_cbor.c (hunk 6 above)
Comment (aggarw13) [reply libraries/standard/serializer/test/unit/iot_tests_serializer_cbor.c]: Done

File: libraries/standard/serializer/src/cbor/iot_serializer_tinycbor_decoder.c (hunk 5 above)
Comment (aggarw13) [reply libraries/standard/serializer/src/cbor/iot_serializer_tinycbor_decoder.c]: Good suggestion. Based on our collective Chime discussion, we decided to remove support of obtaining buffer address and data size for iterator object (as decoder object should be sourced from iterator object through `_get()` operation)

File: libraries/standard/serializer/src/cbor/iot_serializer_tinycbor_decoder.c (hunk 3 above)
Comment (aggarw13) [reply libraries/standard/serializer/src/cbor/iot_serializer_tinycbor_decoder.c]: Okay, have made the change to using assignment copy

File: libraries/standard/serializer/include/iot_serializer.h (hunk 1 above)
Comment (aggarw13) [reply libraries/standard/serializer/include/iot_serializer.h]: Done

**Pull request merged**